feat(aya): Add task storage map type (in the user-space)

vadorovsky opened the pull request:
Task storage is a type of map which uses `task_struct` kernel type as a key. When the task (process) stops, the corresponding entry is automatically removed.

This change add support only in the user-space and tests the functionality with a C program.

Fixes: #203

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/aya-rs/aya/1161)
<!-- Reviewable:end -->

dave-tucker commented on test/integration-test/src/tests/task_storage.rs:
```diff
@@ -0,0 +1,28 @@
+    let store: TaskStorage<_, u32> =
+        TaskStorage::try_from(ebpf.map_mut("task_storage").unwrap()).unwrap();
+
+    let mut child = Command::new("sleep").arg("1").spawn().unwrap();
```
There is no `sleep` binary in our test userspace so this will fail.
You might want to consider using `fork` instead.
https://docs.rs/nix/latest/nix/unistd/fn.fork.html

vadorovsky replied:
Yeah, I noticed. I'm also trying to figure out other options:

* using `thread::spawn`
* using some other LSM hook and looking up for the pid of the testsuite

tamird replied:
Do we need to fork? Can't we just spawn a thread?
<!-- Reviewable comment -OIBnMhqCACbTCrA3hT6:b-m8jl10 -->
<!-- Sent from Reviewable.io -->

vadorovsky replied:
We can't, unfortunately.

Even though in Linux threads are just processes with shared resources, they share PID as well. Threads have the same PID and different TID. Proof:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d65191a69be10325a543cad9c138e9ea

```rust
use std::thread;

fn main() -> anyhow::Result<()> {
    let handles = (0..10).map(|i| thread::spawn(move || {
        let pid = unsafe { libc::getpid() };
        let tid = unsafe { libc::gettid() };
        println!("thread {i}: pid: {pid}, tid: {tid}")
    }));
    
    for handle in handles {
        handle.join().unwrap();
    }
    
    Ok(())
} 
```

```
thread 0: pid: 13, tid: 33
thread 1: pid: 13, tid: 34
thread 2: pid: 13, tid: 35
thread 3: pid: 13, tid: 36
thread 4: pid: 13, tid: 37
thread 5: pid: 13, tid: 38
thread 6: pid: 13, tid: 39
thread 7: pid: 13, tid: 40
thread 8: pid: 13, tid: 41
thread 9: pid: 13, tid: 42
```

The only way to retrieve the entry from a task storage in the user-space is to use the [pidfd](https://www.man7.org/linux/man-pages/man2/pidfd_open.2.html) - a file descriptor corresponding to a process differentiated by PID. There is nothing like that for TID.

I'm glad that you asked that question though, because it made me think of something I didn't consider before. On the eBPF side, my example creates/updates the entry every time we hit `security_task_alloc`, which happens for threads as well. `task_struct` should be unique for threads. Is the task storage on eBPF side smart enough to differentiate threads or does it differentiate only processes? :thinking: Also, if it does differentiate threads on eBPF side, it would mean inconsistency between eBPF and user-space. I need to check...

tamird replied:
Why do you need to differentiate? Just store the *thread* ID in the map.

vadorovsky replied:
You're missing the point how task storage works. You can't just arbitrarily pick the key type.

In the eBPF code, you can only use `task_struct` as a key. [`bpf_task_storage_get`](https://docs.ebpf.io/linux/helper-function/bpf_task_storage_get/) and [`bpf_task_storage_delete`](https://docs.ebpf.io/linux/helper-function/bpf_task_storage_delete/) take a `task_struct` pointer.

But in the user-space, you have to use the pidfd as a key. Nothing else works. See these examples:

https://elixir.bootlin.com/linux/v6.13.1/source/tools/testing/selftests/bpf/map_tests/task_storage_map.c
https://elixir.bootlin.com/linux/v6.13.1/source/tools/testing/selftests/bpf/prog_tests/task_local_storage.c#L283

In all of them, you need to issue the [`pidfd_open`](https://man7.org/linux/man-pages/man2/pidfd_open.2.html) syscall, providing PID as an argument, and then use the returned file descriptor as the key.

Unfortunately, there is no concept of tidfd or anything which would let you use a TID as a key.

tamird replied:
Why can't you store something in the map on thread creation and use pidfd of the process to fetch it from the map?

dave-tucker commented on aya/src/maps/task_storage.rs:
```diff
@@ -0,0 +1,154 @@
+impl<T: Borrow<MapData>, V: Pod> TaskStorage<T, V> {
+    pub(crate) fn new(map: T) -> Result<Self, MapError> {
+        let data = map.borrow();
+        check_kv_size::<c_int, V>(data)?;
```
Other map types use:
```suggestion
        check_kv_size::<u32, V>(data)?;
```

If you can't use u32 - although I'm not sure why not - a comment to explain would be helpful.

dave-tucker commented on aya/src/maps/task_storage.rs:
```diff
@@ -0,0 +1,154 @@
+            TaskStorage::<_, u16>::try_from(&map),
+            Err(MapError::InvalidValueSize {
+                size: 2,
+                expected: 4
```
Coming back to earlier comment - if you're expecting c_int to always be a 4 byte value, you may as well use `u32`.

dave-tucker commented on aya/src/sys/pid_fd.rs:
```diff
@@ -0,0 +1,31 @@
+use libc::pid_t;
+use std::os::fd::{AsRawFd, RawFd};
+/// A file descriptor of a process.
+pub(crate) struct PidFd(RawFd);
```
```suggestion
pub(crate) struct PidFd(MockableFd);
```

or `OwnedFd`. We shouldn't be dealing with `RawFd` directly.

vadorovsky replied:
Done.
<!-- Reviewable comment -OIKI2j16_6TWrnEkKgG:b-896fix -->
<!-- Sent from Reviewable.io -->

dave-tucker commented on aya/src/sys/pid_fd.rs:
```diff
@@ -0,0 +1,31 @@
+use crate::sys::{SysResult, Syscall};
+use libc::pid_t;
+use std::os::fd::{AsRawFd, RawFd};
+/// A file descriptor of a process.
```
You should document that this is already in Rust nightly:
https://doc.rust-lang.org/std/os/linux/process/struct.PidFd.html

We'll want to migrate once it stablizes

vadorovsky replied:
Done.
<!-- Reviewable comment -OIKHWi065kHiOamN2ds:b-896fix -->
<!-- Sent from Reviewable.io -->

dave-tucker commented on aya/src/sys/pid_fd.rs:
```diff
@@ -0,0 +1,31 @@
+impl PidFd {
+    pub(crate) fn open(pid: u32, flags: u32) -> SysResult<Self> {
+        let pid_fd = pidfd_open(pid, flags)?;
+        Ok(Self(pid_fd as i32))
```
Per previous comment, you'll need to create an `OwnedFd` from the syscall.

vadorovsky replied:
Done.
<!-- Reviewable comment -OIKI4PF4LQu8nWDGUBl:b-896fix -->
<!-- Sent from Reviewable.io -->

dave-tucker reviewed (changes requested):
Thanks for this! I know you're still working through the integration test failures but adding some review comments too.

tamird reviewed (changes requested):
> Reviewed 9 of 10 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1161)* status: all files reviewed, 9 unresolved discussions (waiting on @dave-tucker and @vadorovsky)

___
*[`aya/src/sys/pid_fd.rs` line 1 at r2](https://reviewable.io/reviews/aya-rs/aya/1161#-OIBmkBV1BICLdTvxYex:-OIBmkBV1BICLdTvxYey:b-myrmc3) ([raw file](https://github.com/aya-rs/aya/blob/7d45628d442c7eeff6bfc13d0319d947b33c5473/aya/src/sys/pid_fd.rs#L1)):*
> ```XML
> use crate::sys::{SysResult, Syscall};
> ```

do we really need a file for this? seems like overkill to me
___
*[`aya/src/sys/pid_fd.rs` line 12 at r2](https://reviewable.io/reviews/aya-rs/aya/1161#-OIBmg5l1JmTrfM-G3EL:-OIBmg5l1JmTrfM-G3EM:b-dkl64t) ([raw file](https://github.com/aya-rs/aya/blob/7d45628d442c7eeff6bfc13d0319d947b33c5473/aya/src/sys/pid_fd.rs#L12)):*
> ```XML
>     }
> }
> impl AsRawFd for PidFd {
> ```

this is a smell
___
*[`aya/src/sys/pid_fd.rs` line 17 at r2](https://reviewable.io/reviews/aya-rs/aya/1161#-OIBmXaG8v0SRcGWjQBL:-OIBmXaG8v0SRcGWjQBM:b-2ujypz) ([raw file](https://github.com/aya-rs/aya/blob/7d45628d442c7eeff6bfc13d0319d947b33c5473/aya/src/sys/pid_fd.rs#L17)):*
> ```XML
>     }
> }
> impl Drop for PidFd {
> ```

you'd get this for free if you used OwnedFd or MockableFd.


<!-- Sent from Reviewable.io -->

vadorovsky committed:
feat(aya): Add task storage map type (in the user-space)

Task storage is a type of map which uses `task_struct` kernel type as
a key. When the task (process) stops, the corresponding entry is
automatically removed.

This change add support only in the user-space and tests the
functionality with a C program.

Fixes #203

vadorovsky reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1161)* status: 3 of 11 files reviewed, 9 unresolved discussions (waiting on @dave-tucker and @tamird)

___
*[`aya/src/sys/pid_fd.rs` line 1 at r2](https://reviewable.io/reviews/aya-rs/aya/1161#-OIBmkBV1BICLdTvxYex:-OIKGSSL6cukl9_T2F3y:b-svlvr4) ([raw file](https://github.com/aya-rs/aya/blob/7d45628d442c7eeff6bfc13d0319d947b33c5473/aya/src/sys/pid_fd.rs#L1)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

do we really need a file for this? seems like overkill to me
</blockquote></details>

Moved to mod.rs
___
*[`aya/src/sys/pid_fd.rs` line 12 at r2](https://reviewable.io/reviews/aya-rs/aya/1161#-OIBmg5l1JmTrfM-G3EL:-OIKGEBN0bjzK6MEKTz2:b-ouycxg) ([raw file](https://github.com/aya-rs/aya/blob/7d45628d442c7eeff6bfc13d0319d947b33c5473/aya/src/sys/pid_fd.rs#L12)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is a smell
</blockquote></details>

Care to elaborate why? Retrieving the entry from the map requires a raw fd and it's convenient to use that method in `aya/src/maps/task_storage.rs`, line 60, `TaskStorage::get`.
___
*[`aya/src/sys/pid_fd.rs` line 17 at r2](https://reviewable.io/reviews/aya-rs/aya/1161#-OIBmXaG8v0SRcGWjQBL:-OIKJ_HrDmDFicWs1WrF:b-896fix) ([raw file](https://github.com/aya-rs/aya/blob/7d45628d442c7eeff6bfc13d0319d947b33c5473/aya/src/sys/pid_fd.rs#L17)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

you'd get this for free if you used OwnedFd or MockableFd.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tamird reviewed (commented):
> Reviewed 5 of 8 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1161)* status: 10 of 11 files reviewed, 10 unresolved discussions (waiting on @dave-tucker and @vadorovsky)

___
*[`aya/src/sys/mod.rs` line 204 at r4](https://reviewable.io/reviews/aya-rs/aya/1161#-OJre1jz3IfytFlb-67p:-OJre1jz3IfytFlb-67q:bqcdn1c) ([raw file](https://github.com/aya-rs/aya/blob/f9f3c583e9fd0dd5aa96d5f45e3a453c5f9e5ad3/aya/src/sys/mod.rs#L204)):*
> ```XML
> ///
> /// A similar type is provided by the Rust standard library as
> /// [`std::os::linux::process`] as a nigtly-only experimental API. We are
> ```

plz add TODO(https://github.com/rust-lang/rust/issues/82971): ...

but if we really want to use that, we'll need to rework the tests because this API doesn't match `std::os::linux::process::PidFd`
___
*[`aya/src/sys/mod.rs` line 206 at r4](https://reviewable.io/reviews/aya-rs/aya/1161#-OJreKXb8u3BJRyzcAe9:-OJreKXb8u3BJRyzcAeA:bprlnmu) ([raw file](https://github.com/aya-rs/aya/blob/f9f3c583e9fd0dd5aa96d5f45e3a453c5f9e5ad3/aya/src/sys/mod.rs#L206)):*
> ```XML
> /// [`std::os::linux::process`] as a nigtly-only experimental API. We are
> /// planning to migrate to it once it stabilizes.
> pub(crate) struct PidFd(MockableFd);
> ```

you might want to use OwnedFd instead, since we won't be able to inject mockablefd if we use upsream's pidfd
___
*[`test/integration-test/Cargo.toml` line 27 at r4](https://reviewable.io/reviews/aya-rs/aya/1161#-OJrdhweAvXfg3QTO6d-:-OJrdhweAvXfg3QTO6d0:b-p1r1zc) ([raw file](https://github.com/aya-rs/aya/blob/f9f3c583e9fd0dd5aa96d5f45e3a453c5f9e5ad3/test/integration-test/Cargo.toml#L27)):*
> ```TOML
> rand = { workspace = true, features = ["thread_rng"] }
> rbpf = { workspace = true }
> tempfile = { workspace = true }
> ```

where is this used?
___
*[`test/integration-test/src/tests/task_storage.rs` line 21 at r4](https://reviewable.io/reviews/aya-rs/aya/1161#-OJrdRAYEHwkfuQ4s0MF:-OJrdRAYEHwkfuQ4s0MG:b-51mazr) ([raw file](https://github.com/aya-rs/aya/blob/f9f3c583e9fd0dd5aa96d5f45e3a453c5f9e5ad3/test/integration-test/src/tests/task_storage.rs#L21)):*
> ```XML
>     let pid = process::id();
> 
>     sleep(Duration::from_millis(100));
> ```

this is a really long time. maybe we can poll every 1ms up to 100?


<!-- Sent from Reviewable.io -->

vupiggy commented:
@vadorovsky I tried your `TaskStorage` as below:
```Rust
pub struct NetEvent {
    pub saddr: u32,
    pub daddr: u32,
    pub sport: u16,
    pub dport: u16,
}

#[btf_map]
static EVENT_TASKSTORE: TaskStorage<NetEvent> = TaskStorage::new();
```
It compiled, but when I ran the program, I got the below error:
```
Error: error parsing BPF object: BTF error

Caused by:
    0: BTF error
    1: Unexpected BTF type id `63`
``` 
Using `u32` instead of `NetEvent` like `static EVENT_TASKSTORE: TaskStorage<u32> = TaskStorage::new();` works well. Was I missing something?

vadorovsky commented:
We have BTF maps in aya-ebpf now, I will rework this properly with both user-space and BPF support.
